targets/default/jerry-port-default-jobqueue.c

```diff
@@ -0,0 +1,141 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerryscript.h"
+#include "jerry-port.h"
+#include "jerry-port-default.h"
+#include "jmem.h"
+#include "jrt.h"
+
+#ifndef CONFIG_DISABLE_ES2015_PROMISE_BUILTIN
+
+typedef struct jerry_port_queueitem_t jerry_port_queueitem_t;
+
+/**
+ * Description of the queue item.
+ */
+struct jerry_port_queueitem_t
+{
+  jerry_port_queueitem_t *next_p; /**< points to next item */
+  jerry_job_handler_t handler; /**< the handler for the job*/
+  void *job_p; /**< points to the job */
+};
+
+/**
+ * Description of a job queue (FIFO)
+ */
+typedef struct
+{
+  jerry_port_queueitem_t *head_p; /**< points to the head item of the queue */
+  jerry_port_queueitem_t *tail_p; /**< points to the tail item of the queue*/
+} jerry_port_jobqueue_t;
+
+static jerry_port_jobqueue_t queue;
+
+/**
+ * Initialize the job queue
+ */
+void jerry_port_jobqueue_init (void)
+{
+  queue.head_p = NULL;
+  queue.tail_p = NULL;
+} /* jerry_port_jobqueue_init */
+
+/**
+ * Enqueue a job
+ */
+void jerry_port_jobqueue_enqueue (jerry_job_handler_t handler, /**< the handler for the job */
+                                  void *job_p) /**< the job */
+{
+  jerry_port_queueitem_t *item_p = jmem_heap_alloc_block (sizeof (jerry_port_queueitem_t));
+  item_p->job_p = job_p;
+  item_p->handler = handler;
+
+  if (queue.head_p == NULL)
+  {
+    JERRY_ASSERT (queue.tail_p == NULL);
+
+    queue.head_p = item_p;
+    item_p->next_p = NULL;
+    queue.tail_p = item_p;
+
+    return;
+  }
+
+  JERRY_ASSERT (queue.tail_p != NULL);
+
+  queue.tail_p->next_p = item_p;
+  queue.tail_p = item_p;
+} /* jerry_port_jobqueue_enqueue */
+
+/**
+ * Dequeue and get the job.
+ * @return pointer to jerry_port_queueitem_t
+ *         It should be freed with jmem_heap_free_block
+ */
+static jerry_port_queueitem_t *
+jerry_port_jobqueue_dequeue (void)
+{
+  if (queue.head_p == NULL)
+  {
+    JERRY_ASSERT (queue.tail_p == NULL);
+
+    return NULL;
+  }
+
+  JERRY_ASSERT (queue.tail_p != NULL);
+
+  jerry_port_queueitem_t *item_p = queue.head_p;
+  queue.head_p = queue.head_p->next_p;
+
+  return item_p;
+} /* jerry_port_jobqueue_dequeue */
+
+/**
+ * Start the jobqueue.
+ * @return jerry value
+ *         If exception happens in the handler, stop the queue
+ *         and return the exception.
+ *         Otherwise, return undefined
+ */
+jerry_value_t
+jerry_port_jobqueue_run (void)
+{
+  jerry_value_t ret;
+
+  while (true)
+  {
+    jerry_port_queueitem_t *item_p = jerry_port_jobqueue_dequeue ();
+
+    if (item_p == NULL)
+    {
+      return jerry_create_undefined ();
+    }
+
+    void *job_p = item_p->job_p;
+    jerry_job_handler_t handler = item_p->handler;
+    jmem_heap_free_block (item_p, sizeof (jerry_port_queueitem_t));
+    ret = handler (job_p);
+
+    if (jerry_value_has_error_flag (ret))
+    {
+      return ret;
```

Comment:
Reading the *NextJob* section again (http://www.ecma-international.org/ecma-262/6.0/#sec-nextjob-result) I'm not sure whether the loop (recursion in the spec) should be stopped. It doesn't say that it returns after an error result. I guess you could argue that whether or not to return could fall under the implementation specific part.

That said, I think it's more useful to not return (keep looping) and instead call out to another port function to handle the error result, for example by using `jerry_port_console()` to print out the error message. I think something like that is what most people will do in actual projects.

I think it makes sense to expose a `void jerry_port_handle_uncaught_error(jerry_value_t error)` into `jerry-port.h`. Any user binding code that uses async callbacks can then also use the same interface to deal with uncaught errors.